libs3/src/object.c

```diff
@@ -497,8 +502,163 @@ void S3_restore_object(S3BucketContext* bucketContext,
 		restoreObjectCallback,           // fromS3Callback
 		restoreObjectCompleteCallback,   // completeCallback
 		data,                            // callbackData
-		timeoutMs                        // timeoutMs
+		timeoutMs,                       // timeoutMs
+		0                                // xAmzObjectAttributes
 	};
 
 	request_perform(&params, requestContext);
 }
+
+// get object attributes ----------------------------------------------------------------
+typedef struct GetObjectAttributesData
+{
+	SimpleXml simpleXml;
+	S3GetObjectAttributesHandler* handler;
+
+    // 64 is longer than any of these need to be
+    string_buffer(checksumCRC32, 64);
+    string_buffer(checksumCRC32C, 64);
+    string_buffer(checksumCRC64NVME, 64);
+    string_buffer(checksumSHA1, 64);
+    string_buffer(checksumSHA256, 64);
+    string_buffer(checksumType, 64);
+    string_buffer(storageClass, 64);
+    string_buffer(objectSize, 64);
+	void* userdata;
+
+    // Part info not implemented
+} GetObjectAttributesData;
+
+static S3Status GetObjectAttributesCallback(int bufferSize, const char* buffer, void* callbackData)
+{
+	GetObjectAttributesData* goaData = (GetObjectAttributesData*) callbackData;
+	return simplexml_add(&(goaData->simpleXml), buffer, bufferSize);
+}
+
+static void GetObjectAttributesCompleteCallback(S3Status requestStatus,
+                                             const S3ErrorDetails* s3ErrorDetails,
+                                             void* callbackData)
+{
+	GetObjectAttributesData* goaData = (GetObjectAttributesData*) callbackData;
+
+	if (goaData->handler->responseHandler.completeCallback) {
+		(*goaData->handler->responseHandler.completeCallback)(requestStatus, s3ErrorDetails, goaData->userdata);
+	}
+
+	if (goaData->handler->responseXmlCallback) {
+		(*goaData->handler->responseXmlCallback)(goaData->checksumCRC32,
+                goaData->checksumCRC32C,
+                goaData->checksumCRC64NVME,
+                goaData->checksumSHA1,
+                goaData->checksumSHA256,
+                goaData->checksumType,
+                goaData->storageClass,
+                goaData->objectSize,
+                goaData->userdata);
+	}
+
+	simplexml_deinitialize(&(goaData->simpleXml));
+	free(goaData);
+}
+
+static S3Status getObjectAttributesXmlCallback(const char* elementPath, const char* data, int dataLen, void* callbackData)
+{
+	GetObjectAttributesData* goaData = (GetObjectAttributesData*) callbackData;
+	int fit;
+	if (!data) {
+		return S3StatusOK;
+	}
+
+	if (!strcmp(elementPath, "GetObjectAttributesResponse/Checksum/ChecksumCRC32")) {
```

Comment:
Prefer checking the result of `strcmp` with integers. Using negation is not as clear and can cause misinterpretation if the reader isn't paying close attention.